test/e2e/federation-replicaset.go

```diff
@@ -94,32 +98,144 @@ var _ = framework.KubeDescribe("Federation replicasets [Feature:Federation]", fu
+				createReplicaSetOrFail(f.FederationClientset_1_4, rs)
+				defer cleanupFn(rs)
+
+				waitForReplicaSetOrFail(f.FederationClientset_1_4, nsName, rs.Name, clusters, expect)
+				By(fmt.Sprintf("Successfuly created and synced replicaset %q/%q (%v/%v) to clusters", nsName, rs.Name, *rs.Spec.Replicas, rs.Status.Replicas))
+
+				rs = newReplicaSet(nsName, FederationReplicaSetName, 0, pref)
+				updateReplicaSetOrFail(f.FederationClientset_1_4, rs)
+				waitForReplicaSetOrFail(f.FederationClientset_1_4, nsName, rs.Name, clusters, nil)
+				By(fmt.Sprintf("Successfuly updated and synced replicaset %q/%q (%v/%v) to clusters", nsName, rs.Name, *rs.Spec.Replicas, rs.Status.Replicas))
+
+				rs = newReplicaSet(nsName, FederationReplicaSetName, replicas, pref)
+				updateReplicaSetOrFail(f.FederationClientset_1_4, rs)
+				waitForReplicaSetOrFail(f.FederationClientset_1_4, nsName, rs.Name, clusters, expect)
+				By(fmt.Sprintf("Successfuly updated and synced replicaset %q/%q (%v/%v) to clusters", nsName, rs.Name, *rs.Spec.Replicas, rs.Status.Replicas))
+			}
+			createAndUpdateFn(generageFedRsPrefsWithWeight(clusters))
+			createAndUpdateFn(generageFedRsPrefsWithMin(clusters))
+			createAndUpdateFn(generageFedRsPrefsWithMax(clusters))
 		})
 	})
 })
 
-func waitForReplicaSetOrFail(c *federation_release_1_4.Clientset, namespace string, replicaSetName string, clusters map[string]*cluster) {
-	err := waitForReplicaSet(c, namespace, replicaSetName, clusters)
+func generageFedRsPrefsWithWeight(clusters map[string]*cluster) (pref *federation.FederatedReplicaSetPreferences, replicas int32, expect map[string]int32) {
+	clusterNames := make([]string, 0, len(clusters))
+	for clusterName := range clusters {
+		clusterNames = append(clusterNames, clusterName)
+	}
+
+	pref = &federation.FederatedReplicaSetPreferences{
+		Clusters: map[string]federation.ClusterReplicaSetPreferences{},
+	}
+	replicas = 0
+	expect = map[string]int32{}
+
+	for i, clusterName := range clusterNames {
+		if i != 0 {
+			clusterRsPref := pref.Clusters[clusterName]
+			clusterRsPref.Weight = int64(i)
+			replicas += int32(i)
+			expect[clusterName] = int32(i)
+		}
+	}
+
+	return
+}
+
+func generageFedRsPrefsWithMin(clusters map[string]*cluster) (pref *federation.FederatedReplicaSetPreferences, replicas int32, expect map[string]int32) {
+	clusterNames := make([]string, 0, len(clusters))
+	for clusterName := range clusters {
+		clusterNames = append(clusterNames, clusterName)
+	}
+
+	pref = &federation.FederatedReplicaSetPreferences{
+		Clusters: map[string]federation.ClusterReplicaSetPreferences{
+			clusterNames[0]: {Weight: 100},
+		},
+	}
+	replicas = 0
+	expect = map[string]int32{}
+
+	for i, clusterName := range clusterNames {
+		if i != 0 {
+			clusterRsPref := pref.Clusters[clusterName]
+			clusterRsPref.Weight = int64(1)
+			clusterRsPref.MinReplicas = int64(i + 2)
+			replicas += int32(i + 2)
+			expect[clusterName] = int32(i + 2)
+		}
+	}
+	replicas += 1
```

Comment:
Add a comment saying the extra replica should go to the 0th cluster (since it has the highest weight)